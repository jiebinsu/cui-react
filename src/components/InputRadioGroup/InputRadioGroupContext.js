import { createContext, useContext } from "react";

export const RadioContext = createContext({
  inGroup: false,
});

export const useRadioContext = () => useContext(RadioContext);
