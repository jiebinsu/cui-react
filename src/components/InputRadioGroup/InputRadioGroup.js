import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { RadioContext } from "./InputRadioGroupContext";
import "./inputRadioGroup.scss";

const InputRadioGroup = ({
  id,
  legend,
  hint,
  hasError,
  errorMsg,
  className,
  initialValue,
  value,
  children,
  onChange,
  ...props
}) => {
  const [checked, setChecked] = useState(initialValue);

  const updateState = (nextValue) => {
    setChecked(nextValue);
    if (onChange) onChange(nextValue);
  };

  const providerValue = {
    updateState,
    currentState: checked,
    inGroup: true,
  };

  useEffect(() => {
    if (value === undefined) return;
    setChecked(value);
  }, [value]);

  const hintId = `${id}-hint`;
  const errorId = `${id}-error`;
  const classNames = ["coop-c-form-choice"];
  const ariaDescribedBy = [];

  if (hint) ariaDescribedBy.push(hintId);

  if (hasError) {
    classNames.push("coop-form__invalid");
    ariaDescribedBy.push(errorId);
  }

  classNames.push(className);

  return (
    <RadioContext.Provider value={providerValue}>
      <div className="coop-form__row">
        <fieldset
          className={classNames.join(" ")}
          aria-describedby={ariaDescribedBy && ariaDescribedBy.join(" ")}
          {...props}
        >
          <legend className="coop-form__legend coop-c-form-choice__legend">
            {legend}
          </legend>
          {hint && (
            <p id={`${id}-hint`} className="coop-form__hint">
              {hint}
            </p>
          )}
          {hasError && (
            <p id={`${id}-error`} className="coop-form__error">
              {errorMsg}
            </p>
          )}
          {children}
        </fieldset>
      </div>
    </RadioContext.Provider>
  );
};

export default InputRadioGroup;

InputRadioGroup.defaultProps = {
  hasError: false,
  hint: null,
  errorMsg: null,
  children: null,
  className: null,
  initialValue: null,
  value: null,
  onChange: null,
};

InputRadioGroup.propTypes = {
  id: PropTypes.string.isRequired,
  legend: PropTypes.string.isRequired,
  hint: PropTypes.string,
  hasError: PropTypes.bool,
  errorMsg: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  initialValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  onChange: PropTypes.func,
};
