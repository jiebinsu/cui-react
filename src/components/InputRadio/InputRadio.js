import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { useRadioContext } from "../InputRadioGroup/InputRadioGroupContext";
import classNames from "../../utils/classNames";
import "./inputRadio.scss";

const InputRadio = ({
  id,
  name,
  value,
  className,
  label,
  checked,
  value: radioValue,
  disabled,
  onChange,
  ...props
}) => {
  const [selfChecked, setSelfChecked] = useState(!!checked);
  const context = useRadioContext();
  const { currentState, updateState, inGroup } = context;
  const classes = classNames(
    "coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button",
    [className]
  );

  if (inGroup) {
    useEffect(() => {
      setSelfChecked(currentState === radioValue);
    }, [currentState, radioValue]);
  }

  const onChangeHandler = (event) => {
    if (disabled) return;
    const selfEvent = {
      target: {
        checked: !selfChecked,
      },
      stopPropagation: event.stopPropagation,
      preventDefault: event.preventDefault,
      nativeEvent: event,
    };
    setSelfChecked(!selfChecked);
    if (inGroup) {
      if (updateState) updateState(radioValue);
    }
    if (onChange) onChange(selfEvent);
  };

  useEffect(() => {
    if (checked === undefined) return;
    setSelfChecked(Boolean(checked));
  }, [checked]);

  const tagAttributes = {
    id,
    name,
    value: radioValue,
    label,
    disabled,
    type: "radio",
    className: classes,
    checked: selfChecked,
    onChange: onChangeHandler,
    ...props,
  };

  return (
    <div className="coop-c-form-choice">
      <input {...tagAttributes} />
      <label
        className="coop-form__label coop-c-form-choice__label"
        htmlFor={id}
      >
        {label}
      </label>
    </div>
  );
};

InputRadio.defaultProps = {
  disabled: false,
  className: null,
  checked: null,
  onChange: null,
};

InputRadio.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  className: PropTypes.string,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};

export default InputRadio;
