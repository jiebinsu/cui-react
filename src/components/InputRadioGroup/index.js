import InputRadioGroup from "./InputRadioGroup";

export default InputRadioGroup;
