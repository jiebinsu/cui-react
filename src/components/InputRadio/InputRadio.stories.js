import React from "react";
import InputRadio from "./InputRadio";

export default {
  title: "Base/InputRadio",
  component: InputRadio,
};

export const Basic = (args) => <InputRadio {...args} />;
Basic.args = {
  id: "basic",
  name: "basic",
  label: "Basic Radio",
  value: 1,
};
