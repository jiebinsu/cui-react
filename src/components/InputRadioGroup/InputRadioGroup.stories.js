import React from "react";
import InputRadioGroup from "./InputRadioGroup";
import InputRadio from "../InputRadio";

export default {
  title: "Base/InputRadioGroup",
  component: InputRadioGroup,
};

export const RadioExample = () => (
  <>
    <InputRadioGroup
      id="group-1"
      legend="Radio Group with hint"
      hint="Hint message"
    >
      <InputRadio id="radio-1" name="radio-1" label="Radio 1" value={1} />
      <InputRadio id="radio-2" name="radio-2" label="Radio 2" value={2} />
      <InputRadio id="radio-3" name="radio-3" label="Radio 3" value={3} />
    </InputRadioGroup>
    <InputRadioGroup
      id="group-2"
      legend="Radio Group with hint and error"
      hint="Hint message"
      errorMsg="Error message"
      hasError
    >
      <InputRadio id="radio-4" name="radio-4" label="Radio 4" value={4} />
      <InputRadio id="radio-5" name="radio-5" label="Radio 5" value={5} />
      <InputRadio id="radio-6" name="radio-6" label="Radio 6" value={6} />
    </InputRadioGroup>
  </>
);
