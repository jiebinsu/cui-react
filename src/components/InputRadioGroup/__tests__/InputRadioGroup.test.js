import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import InputRadioGroup from "../index";
import InputRadio from "../../InputRadio";

describe("InputRadioGroup", () => {
  it("should render correctly", () => {
    const wrapper = mount(
      <InputRadioGroup
        id="group-1"
        legend="Radio Group with hint"
        hint="Hint message"
      />
    );
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should support hints and errors", () => {
    const All = () => (
      <div>
        <InputRadioGroup
          id="group-1"
          legend="Radio Group with hint"
          hint="Hint message"
        >
          <InputRadio id="radio-1" name="radio-1" label="Radio 1" value={1} />
          <InputRadio id="radio-2" name="radio-2" label="Radio 2" value={2} />
          <InputRadio id="radio-3" name="radio-3" label="Radio 3" value={3} />
        </InputRadioGroup>
        <InputRadioGroup
          id="group-2"
          legend="Radio Group with hint and error"
          hint="Hint message"
          errorMsg="Error message"
          hasError
        >
          <InputRadio id="radio-4" name="radio-4" label="Radio 4" value={4} />
          <InputRadio id="radio-5" name="radio-5" label="Radio 5" value={5} />
          <InputRadio id="radio-6" name="radio-6" label="Radio 6" value={6} />
        </InputRadioGroup>
      </div>
    );

    const tree = renderer.create(<All />).toJSON();
    expect(tree).toMatchSnapshot();

    const wrapper = mount(<All />);
    expect(() => wrapper.unmount()).not.toThrow();
  });
});
