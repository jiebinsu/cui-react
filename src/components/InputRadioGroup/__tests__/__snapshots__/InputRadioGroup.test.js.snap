// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRadioGroup should support hints and errors 1`] = `
<div>
  <div
    className="coop-form__row"
  >
    <fieldset
      aria-describedby="group-1-hint"
      className="coop-c-form-choice "
    >
      <legend
        className="coop-form__legend coop-c-form-choice__legend"
      >
        Radio Group with hint
      </legend>
      <p
        className="coop-form__hint"
        id="group-1-hint"
      >
        Hint message
      </p>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button"
          disabled={false}
          id="radio-1"
          name="radio-1"
          onChange={[Function]}
          type="radio"
          value={1}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="radio-1"
        >
          Radio 1
        </label>
      </div>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button"
          disabled={false}
          id="radio-2"
          name="radio-2"
          onChange={[Function]}
          type="radio"
          value={2}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="radio-2"
        >
          Radio 2
        </label>
      </div>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button"
          disabled={false}
          id="radio-3"
          name="radio-3"
          onChange={[Function]}
          type="radio"
          value={3}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="radio-3"
        >
          Radio 3
        </label>
      </div>
    </fieldset>
  </div>
  <div
    className="coop-form__row"
  >
    <fieldset
      aria-describedby="group-2-hint group-2-error"
      className="coop-c-form-choice coop-form__invalid "
    >
      <legend
        className="coop-form__legend coop-c-form-choice__legend"
      >
        Radio Group with hint and error
      </legend>
      <p
        className="coop-form__hint"
        id="group-2-hint"
      >
        Hint message
      </p>
      <p
        className="coop-form__error"
        id="group-2-error"
      >
        Error message
      </p>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button"
          disabled={false}
          id="radio-4"
          name="radio-4"
          onChange={[Function]}
          type="radio"
          value={4}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="radio-4"
        >
          Radio 4
        </label>
      </div>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button"
          disabled={false}
          id="radio-5"
          name="radio-5"
          onChange={[Function]}
          type="radio"
          value={5}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="radio-5"
        >
          Radio 5
        </label>
      </div>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button"
          disabled={false}
          id="radio-6"
          name="radio-6"
          onChange={[Function]}
          type="radio"
          value={6}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="radio-6"
        >
          Radio 6
        </label>
      </div>
    </fieldset>
  </div>
</div>
`;
